From 50802af3763ad8ba78a39bb8f4722c406e229475 Mon Sep 17 00:00:00 2001
From: Frank Wang <frank.wang@rock-chips.com>
Date: Tue, 10 Oct 2017 10:19:22 +0800
Subject: [PATCH] phy: rockchip-inno-usb3: fixed usb3 devices detected failed

We found the usb-phy lost devices detected ability after continuously
disconnect/connect, this patch add usb2-phy reset/deassert as one part
of previous workaround (commit c1ebf31) to fix it.

Change-Id: Ib7112047eb0f5030406389aa9c8ebd599f8118be
Signed-off-by: Frank Wang <frank.wang@rock-chips.com>
---
 drivers/phy/rockchip/phy-rockchip-inno-usb3.c | 30 +++++++++----------
 1 file changed, 15 insertions(+), 15 deletions(-)

diff --git a/drivers/phy/rockchip/phy-rockchip-inno-usb3.c b/drivers/phy/rockchip/phy-rockchip-inno-usb3.c
index 961942d7a562..22bd72ab4383 100644
--- a/drivers/phy/rockchip/phy-rockchip-inno-usb3.c
+++ b/drivers/phy/rockchip/phy-rockchip-inno-usb3.c
@@ -60,12 +60,12 @@ enum rockchip_u3phy_pipe_pwr {
 };
 
 enum rockchip_u3phy_rest_req {
-	U2_POR_RSTN	= 0,
-	U3_POR_RSTN	= 1,
+	U3_POR_RSTN	= 0,
+	U2_POR_RSTN	= 1,
 	PIPE_MAC_RSTN	= 2,
 	UTMI_MAC_RSTN	= 3,
-	UTMI_APB_RSTN	= 4,
-	PIPE_APB_RSTN	= 5,
+	PIPE_APB_RSTN	= 4,
+	UTMI_APB_RSTN	= 5,
 	U3PHY_RESET_MAX	= 6,
 };
 
@@ -328,23 +328,23 @@ static void rockchip_u3phy_rest_deassert(struct rockchip_u3phy *u3phy,
 
 	if (flag & U3PHY_APB_RST) {
 		dev_dbg(u3phy->dev, "deassert APB bus interface reset\n");
-		for (rst = UTMI_APB_RSTN; rst <= PIPE_APB_RSTN; rst++) {
+		for (rst = PIPE_APB_RSTN; rst <= UTMI_APB_RSTN; rst++) {
 			if (u3phy->rsts[rst])
 				reset_control_deassert(u3phy->rsts[rst]);
 		}
 	}
 
 	if (flag & U3PHY_POR_RST) {
-		usleep_range(15, 20);
+		usleep_range(12, 15);
 		dev_dbg(u3phy->dev, "deassert u2 and u3 phy power on reset\n");
-		for (rst = U2_POR_RSTN; rst <= U3_POR_RSTN; rst++) {
+		for (rst = U3_POR_RSTN; rst <= U2_POR_RSTN; rst++) {
 			if (u3phy->rsts[rst])
 				reset_control_deassert(u3phy->rsts[rst]);
 		}
 	}
 
 	if (flag & U3PHY_MAC_RST) {
-		usleep_range(2000, 2100);
+		usleep_range(1200, 1500);
 		dev_dbg(u3phy->dev, "deassert pipe and utmi MAC reset\n");
 		for (rst = PIPE_MAC_RSTN; rst <= UTMI_MAC_RSTN; rst++)
 			if (u3phy->rsts[rst])
@@ -812,10 +812,7 @@ static int rockchip_u3phy_on_init(struct usb_phy *usb_phy)
 	struct rockchip_u3phy *u3phy =
 		container_of(usb_phy, struct rockchip_u3phy, usb_phy);
 
-	reset_control_deassert(u3phy->rsts[U3_POR_RSTN]);
-	usleep_range(250, 300);
-	reset_control_deassert(u3phy->rsts[PIPE_MAC_RSTN]);
-
+	rockchip_u3phy_rest_deassert(u3phy, U3PHY_POR_RST | U3PHY_MAC_RST);
 	return 0;
 }
 
@@ -823,10 +820,13 @@ static void rockchip_u3phy_on_shutdown(struct usb_phy *usb_phy)
 {
 	struct rockchip_u3phy *u3phy =
 		container_of(usb_phy, struct rockchip_u3phy, usb_phy);
+	int rst;
 
-	reset_control_assert(u3phy->rsts[U3_POR_RSTN]);
-	reset_control_assert(u3phy->rsts[PIPE_MAC_RSTN]);
-	usleep_range(15, 20);
+	for (rst = 0; rst < U3PHY_RESET_MAX; rst++)
+		if (u3phy->rsts[rst] && rst != UTMI_APB_RSTN &&
+		    rst != PIPE_APB_RSTN)
+			reset_control_assert(u3phy->rsts[rst]);
+	udelay(1);
 }
 
 static int rockchip_u3phy_on_disconnect(struct usb_phy *usb_phy,
-- 
2.35.3

